### PR TITLE
[opt] Add regularation and Nesterov for mergerd_momentum op

### PR DESCRIPTION
**Pull request by zhangbo9674**
### PR types
Performance optimization

### PR changes
OPs
### Describe
增强mergerd_momentum op功能，包括：
- 由仅支持输入单个lr到支持输入多lrs（数量与输入的参数一致）；
- 添加use_nesterov属性，支持use_nesterov策略的计算；
- 添加regularization属性，支持regularization计算。

### CODE REVIEW

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 1]
```diff
@@ -147,20 +148,45 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+      PADDLE_ENFORCE_EQ(n, regularization_methods.size(),
+                        platform::errors::InvalidArgument(
+                            "Attr(regularization_method) number must be equal "
+                            "to Input(Param) number."));
+      PADDLE_ENFORCE_EQ(n, regularization_coeffs.size(),
+                        platform::errors::InvalidArgument(
+                            "Attr(regularization_coeff) number must be equal "
+                            "to Input(Param) number."));
+    }
+    VLOG(1) << use_nesterov << regularization_methods.size()
```
Comment (zhiqiu) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h]:
```suggestion
    VLOG(5) << "use_nesterov: " << use_nesterov <<",  regularization_methods.size(): " <<  regularization_methods.size()
```

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 2]
```diff
@@ -147,20 +148,45 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+    auto regularization_coeffs =
+        ctx.Attr<std::vector<float>>("regularization_coeff");
+    if (regularization_methods.size() != 0) {
+      PADDLE_ENFORCE_EQ(n, regularization_methods.size(),
+                        platform::errors::InvalidArgument(
+                            "Attr(regularization_method) number must be equal "
+                            "to Input(Param) number."));
+      PADDLE_ENFORCE_EQ(n, regularization_coeffs.size(),
+                        platform::errors::InvalidArgument(
+                            "Attr(regularization_coeff) number must be equal "
```
Comment (zhiqiu) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h]: The size of Attr(regularization_coeff) must be equal to the size of Input(Param), but got the size of Attr(regularization_coeff) is %d, the size of Input(Param) is %d

Comment (zhiqiu) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: Try to  make the error message helpful, same for others.

File: paddle/fluid/operators/optimizers/merged_momentum_op.cc [hunk 3]
```diff
@@ -68,6 +69,18 @@ class MergedMomentumOpMaker : public framework::OpProtoAndCheckerMaker {
         .AsDispensable()
         .AsDuplicable();
     AddAttr<float>("mu", "(float) Momentum coefficient");
+    AddAttr<bool>("use_nesterov",
+                  "(bool, default false) "
+                  "Use Nesterov Momentum")
```
Comment (zhiqiu) [inline paddle/fluid/operators/optimizers/merged_momentum_op.cc]:
```suggestion
                  "Use Nesterov Momentum or not")
```

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 4]
```diff
@@ -182,14 +208,82 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+        PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL(true);
+      } else {
+        PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL(false);
+      }
+#undef PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL
     } else {
-      PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL(false);
-    }
+      for (size_t idx = 0; idx < n; idx++) {
+        std::string regularization_method = " ";
```
Comment (zhiqiu) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h]:
```suggestion
        std::string regularization_method = "";
```

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 1 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: Done, tkx!

File: paddle/fluid/operators/optimizers/merged_momentum_op.cc (hunk 3 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.cc]: Done, tkx!

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 4 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: Done, tkx!

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 2 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: Done, tks!

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 5]
```diff
@@ -182,14 +240,82 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+        MPType regularization_coeff = static_cast<MPType>(0.0);
+        if (regularization_coeffs.size() != 0) {
+          regularization_coeff =
+              static_cast<MPType>(regularization_coeffs[idx]);
+        }
+        auto lr_temp = lrs.size() > 1 ? lrs[idx] : lrs[0];
 
-#undef PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL
+        params_out[idx]->data<T>();
+        velocitys_out[idx]->data<MPType>();
```
Comment (sneaxiy) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h]: I do not know what is the purpose to write these 2 lines? Just check whether `params_out[idx]` and `velocitys_out[idx]` is properly initialized?

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 6]
```diff
@@ -182,14 +240,82 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+#undef PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL
     } else {
-      PADDLE_LAUNCH_MERGED_MOMENTUM_KERNEL(false);
-    }
+      for (size_t idx = 0; idx < n; idx++) {
+        std::string regularization_method = "";
+        if (regularization_methods.size() != 0) {
+          regularization_method = regularization_methods[idx];
+        }
+        RegularizationType regularization_flag{RegularizationType::kNONE};
```
Comment (sneaxiy) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h]: NIT. Not required to change. Maybe the following code would be simpler:
```c++
RegularizationType regularization_flag = regularization_methods.size() > 0 &&  regularization_methods[idx] == "l2_decay" ? RegularizationType::kL2DECAY : RegularizationType::kNONE.
```

File: paddle/fluid/operators/optimizers/merged_momentum_op.h [hunk 7]
```diff
@@ -182,14 +240,82 @@ class MergedMomentumOpKernel : public framework::OpKernel<T> {
...
+              VLOG(10)
+                  << "Launch MergedMomentum gpu kernel no_nesterov kL2DECAY.";
+            } else {
+              PADDLE_LAUNCH_DENSE_MTMOMENTUM_KERNEL(NoNesterov,
+                                                    RegularizationType::kNONE);
+              VLOG(10) << "Launch MergedMomentum gpu kernel no_nesterov kNONE.";
+            }
+          }
+        }
+      }
```
Comment (sneaxiy) [inline paddle/fluid/operators/optimizers/merged_momentum_op.h:311]: Seems too many duplicate codes with `momentum_op.h`. Maybe we can use a common function defined in `momentum_op.h`?

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 5 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: tks, this code has been deleted.

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 6 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h]: tks, this code has been modified according to the comments.

File: paddle/fluid/operators/optimizers/merged_momentum_op.h (hunk 7 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/optimizers/merged_momentum_op.h:311]: I think these codes have reused the `DenseMomentumFunctor` function in `momentum_op.h`.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**